src/Data/Array/Accelerate/Debug/Internal/Profile.hs

```diff
@@ -0,0 +1,172 @@
+
+#ifdef ACCELERATE_DEBUG
+import Control.Monad
+#endif
+#ifdef ACCELERATE_PROFILE
+import qualified Data.Array.Accelerate.Debug.Internal.Tracy         as Tracy
+#endif
+
+import Data.Atomic                                                  ( Atomic )
+import qualified Data.Atomic                                        as Atomic
+
+import Data.Char
+import Foreign.C.String
+import Foreign.Ptr
+import Language.Haskell.TH
+import Language.Haskell.TH.Syntax
+
+import GHC.Ptr
+
+
+-- Embed some string data into the constant section and grab a pointer
+-- directly to it.
+--
+-- XXX: This only allows us to track a single nursery, but in reality it is
+-- one per remote memory space. May or may not be useful to separate this.
+--
+runQ $ sequence
+  [ sigD (mkName "___nursery") (conT ''CString)
+  , valD (varP (mkName "___nursery")) (normalB (conE 'Ptr `appE` litE (stringPrimL (map (fromIntegral . ord) "nursery\0")))) []
+  ]
+
+-- Allocations in the local memory space
+--
+{-# INLINE local_memory_alloc #-}
+{-# INLINE local_memory_free  #-}
+local_memory_alloc :: Ptr a -> Int -> IO ()
+local_memory_free  :: Ptr a -> IO ()
+#ifndef ACCELERATE_DEBUG
+local_memory_alloc _ _ = return ()
+local_memory_free _    = return ()
+#else
+
+local_memory_alloc _p n = do
+#ifdef ACCELERATE_PROFILE
+  Tracy.emit_memory_alloc _p (fromIntegral n) 0
+#endif
+  void $ Atomic.add __total_bytes_allocated_local (fromIntegral n)
+
+local_memory_free _p = do
+#ifdef ACCELERATE_PROFILE
+  Tracy.emit_memory_free _p 0
+#endif
+  return ()
+#endif
+
+
+-- Allocations in the remote memory space
+--
+{-# INLINE remote_memory_alloc #-}
+{-# INLINE remote_memory_free  #-}
+{-# INLINE remote_memory_evict #-}
+remote_memory_alloc :: CString -> Ptr a -> Int -> IO ()
+remote_memory_free  :: CString -> Ptr a -> IO ()
+remote_memory_evict :: CString -> Ptr a -> Int -> IO ()
+#ifndef ACCELERATE_DEBUG
+remote_memory_alloc _ _ _ = return ()
+remote_memory_free _ _    = return ()
+remote_memory_evict _ _ _ = return ()
+#else
+remote_memory_alloc _name _ptr bytes = do
+#ifdef ACCELERATE_PROFILE
+  Tracy.emit_memory_alloc_named _ptr (fromIntegral bytes) 0 _name
+#endif
+  void $ Atomic.add __total_bytes_allocated_remote (fromIntegral bytes)
+
+remote_memory_free _name _ptr = do
+#ifdef ACCELERATE_PROFILE
+  Tracy.emit_memory_free_named _ptr 0 _name
+#endif
+  return ()
+
+remote_memory_evict name ptr bytes = do
+  void $ Atomic.add __num_evictions 1
+  void $ Atomic.add __total_bytes_evicted_from_remote (fromIntegral bytes)
+  remote_memory_free name ptr
+  memcpy_from_remote bytes
+#endif
+
+remote_memory_alloc_nursery :: Ptr a -> Int -> IO ()
+remote_memory_free_nursery  :: Ptr a -> IO ()
+#ifndef ACCELERATE_PROFILE
+remote_memory_alloc_nursery _ _ = return ()
+remote_memory_free_nursery _    = return ()
+#else
+remote_memory_alloc_nursery p n = Tracy.emit_memory_alloc_named p (fromIntegral n) 0 ___nursery
+remote_memory_free_nursery p    = Tracy.emit_memory_free_named p 0 ___nursery
+#endif
```

Comment:
There is no way to instrument the GC with these, is there? That would make it much easier to spot when the GC is causing delays. So similar to how you're able to override `operator new/delete` in C++ and use malloc/free hooks in (GNU) C.

Reply:
Not that I know of but that would be useful. There seems to be a [`gcDoneHook`](https://github.com/ghc/ghc/blob/5abf59976c7335df760e5d8609d9488489478173/rts/RtsFlags.c#L78) in the RTS that we might be able to use, but that is probably obvious from the memory usage graph already.